src/assets-loader/assets-loader.ts

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { Injectable } from '@nestjs/common';
+import { AssetDescriptionType, Transaction } from '@prisma/client';
+import { AssetDescriptionsService } from '../asset-descriptions/asset-descriptions.service';
+import { AssetsService } from '../assets/assets.service';
+import { BlockDto } from '../blocks/dto/upsert-blocks.dto';
+import { BlockOperation } from '../blocks/enums/block-operation';
+import { LoggerService } from '../logger/logger.service';
+import { BasePrismaClient } from '../prisma/types/base-prisma-client';
+import { TransactionDto } from '../transactions/dto/upsert-transactions.dto';
+import { TransactionsService } from '../transactions/transactions.service';
+
+@Injectable()
+export class AssetsLoader {
+  constructor(
+    private readonly assetsService: AssetsService,
+    private readonly assetDescriptionsService: AssetDescriptionsService,
+    private readonly logger: LoggerService,
+    private readonly transactionsService: TransactionsService,
+  ) {}
+
+  async loadDescriptions(
+    block: BlockDto,
+    prisma: BasePrismaClient,
+  ): Promise<void> {
+    const main = block.type === BlockOperation.CONNECTED;
+
+    for (const dto of block.transactions) {
+      const transaction = await this.transactionsService.findByHashOrThrow(
+        dto.hash,
+        prisma,
+      );
+
+      if (main) {
+        await this.createAssetDescriptions(dto, transaction, prisma);
+      } else {
+        await this.deleteAssetDescriptions(transaction, prisma);
+      }
+    }
+  }
+
+  private async createAssetDescriptions(
+    dto: TransactionDto,
+    transaction: Transaction,
+    prisma: BasePrismaClient,
+  ): Promise<void> {
+    const existingDescriptions =
+      await this.assetDescriptionsService.findByTransaction(
+        transaction,
+        prisma,
+      );
+    if (existingDescriptions.length > 0) {
+      this.logger.debug(
+        `Skipping description loading for transaction '${transaction.hash}'`,
+      );
+      return;
+    }
+
+    for (const mint of dto.mints) {
+      const asset = await this.assetsService.upsert(
+        {
+          identifier: mint.id,
+          metadata: mint.metadata,
+          name: mint.name,
+          owner: mint.owner,
+        },
+        transaction,
+        prisma,
+      );
+
```

Comment:
What if we replay these, would there be new descriptions created? Should we delete the existing entries for a `transaction` before we create the new ones? We would also need to retroactively update the supply for the duplicates right?

Reason I ask is because it turned out to be very useful for `multi-asset` stuff to just replay the syncer from the beginning of time idempotently.

Reply:
I am assuming that if descriptions exist for a transaction hash, I'll skip processing [here](https://github.com/iron-fish/ironfish-api/pull/1230/files#diff-92a486d7d498d7daa503db784be52172c0e67eed3f0868e3e3bbf54271458afaR49-R59). I can change this to delete and then re-create. What do you think?

Reply:
Yea I think changing to delete/recreate is safer for replay, wouldn't require any manual modifications to DB.